pkg/entrypoint/admissioncontroller/podwhitelist.go

```diff
@@ -0,0 +1,145 @@
+package admissioncontroller
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"strings"
+
+	_ "github.com/openshift/origin/pkg/api/install"
+	oapps "github.com/openshift/origin/pkg/apps/apis/apps"
+	"github.com/openshift/origin/pkg/security/apiserver/securitycontextconstraints"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/apis/batch"
+	"k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+)
+
+func (ac *admissionController) handleWhitelist(w http.ResponseWriter, r *http.Request) {
+	unpackers := map[string]func(runtime.Object) (core.PodSpec, metav1.ObjectMeta, string){
+		"Pod": func(o runtime.Object) (core.PodSpec, metav1.ObjectMeta, string) {
+			pod := o.(*core.Pod)
+			return pod.Spec, pod.ObjectMeta, pod.Namespace
+		},
+		"DaemonSet": func(o runtime.Object) (core.PodSpec, metav1.ObjectMeta, string) {
+			ds := o.(*extensions.DaemonSet)
+			return ds.Spec.Template.Spec, ds.Spec.Template.ObjectMeta, ds.Namespace
+		},
+		"ReplicaSet": func(o runtime.Object) (core.PodSpec, metav1.ObjectMeta, string) {
+			rs := o.(*extensions.ReplicaSet)
+			return rs.Spec.Template.Spec, rs.Spec.Template.ObjectMeta, rs.Namespace
+		},
+		"StatefulSet": func(o runtime.Object) (core.PodSpec, metav1.ObjectMeta, string) {
+			ss := o.(*apps.StatefulSet)
+			return ss.Spec.Template.Spec, ss.Spec.Template.ObjectMeta, ss.Namespace
+		},
+		"Job": func(o runtime.Object) (core.PodSpec, metav1.ObjectMeta, string) {
+			job := o.(*batch.Job)
+			return job.Spec.Template.Spec, job.Spec.Template.ObjectMeta, job.Namespace
+		},
+		"CronJob": func(o runtime.Object) (core.PodSpec, metav1.ObjectMeta, string) {
+			cj := o.(*batch.CronJob)
+			return cj.Spec.JobTemplate.Spec.Template.Spec, cj.Spec.JobTemplate.Spec.Template.ObjectMeta, cj.Namespace
+		},
+		"DeploymentConfig": func(o runtime.Object) (core.PodSpec, metav1.ObjectMeta, string) {
+			dc := o.(*oapps.DeploymentConfig)
+			return dc.Spec.Template.Spec, dc.Spec.Template.ObjectMeta, dc.Namespace
+		},
+		"Deployment": func(o runtime.Object) (core.PodSpec, metav1.ObjectMeta, string) {
+			dp := o.(*extensions.Deployment)
+			return dp.Spec.Template.Spec, dp.Spec.Template.ObjectMeta, dp.Namespace
+		},
+	}
+	req, errcode := ac.getAdmissionReviewRequest(r)
+	if errcode != 0 {
+		http.Error(w, http.StatusText(errcode), errcode)
+		return
+	}
+	if req.UID == "" || req.Kind.Version == "" || req.Kind.Kind == "" {
+		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+		return
+	}
+	gvk := schema.GroupVersionKind{Group: req.Kind.Group, Version: req.Kind.Version, Kind: req.Kind.Kind}
```

Comment:
think we discussed this before - do you really need this line?

Reply:
Turns out that it's needed because the embedded object doesn't have GroupVersionKind set, so the Decode function has to get that value from the review request.